gapir/cc/in_memory_resource_cache.cpp

```diff
@@ -23,174 +23,237 @@
+      mIDGenerator(0) {}
 
-InMemoryResourceCache::~InMemoryResourceCache() {
-  while (mHead->next != mHead) {
-    destroy(mHead->next);
+InMemoryResourceCache::~InMemoryResourceCache() {}
+
+bool InMemoryResourceCache::putCache(const Resource& res, const void* resData) {
+  if (res.getSize() > mMemoryLimit) {
+    return false;
   }
-  delete mHead;
-}
 
-void InMemoryResourceCache::clear() {
-  mCache.clear();
-  while (mHead->next != mHead) {
-    mHead = destroy(mHead);
+  // If we need to evict anything to get this new entry to fit. Now's the time
+  // to do it.
+  resize(mMemoryLimit - res.getSize());
+
+  // Get a new ID for this cache entry that is larger than any current ID.
+  auto newID = mIDGenerator++;
+
+  // Try to allocate some memory. If we get an allocation faulture, throw more
+  // stuff out until we succeed. This might happen even if we passed the memory
+  // limit check above, because we cannot control the other applications running
+  // on our device and how much memory they might use.
+  MemoryAllocator::Handle newMemory;
+  while (newMemory == nullptr) {
+    newMemory = mAllocator->allocatePurgable(res.getSize());
+
+    if (newMemory == nullptr) {
+      assert(evictLeastRecentlyUsed(mMemoryUse / 2));
+    }
   }
-  *mHead = Block(0, mBufferSize);
-  mHead->next = mHead->prev = mHead;
+
+  assert(mResourceIndex.find(res.getID()) == mResourceIndex.end());
+  assert(mResources.find(newID) == mResources.end());
+
+  // Enter the new allocation into our records.
+  mResourceIndex[res.getID()] = newID;
+  mResources[newID] = std::make_pair(res, newMemory);
+
+  // Add the memory allocated to our record of how much we have "live"
+  mMemoryUse += res.getSize();
+
+  // Copy the bits into the cache.
+  memcpy(&newMemory[0], resData, res.getSize());
+
+  return true;
 }
 
-bool InMemoryResourceCache::resize(size_t newSize) {
-  GAPID_DEBUG("Cache resizing: %zu -> %zu", mBufferSize, newSize);
-  if (newSize == mBufferSize) {
-    return true;  // No change.
-  }
+bool InMemoryResourceCache::hasCache(const Resource& res) {
+  return findCache(res) != mResources.end();
+}
 
-  Block* first = this->first();
-  Block* last = this->last();
+bool InMemoryResourceCache::loadCache(const Resource& res, void* target) {
+  mCacheAccesses++;
 
-  // Remove all the blocks that are completely beyond the end of the new size.
-  while (last != first && last->offset > newSize) {
-    last = last->prev;
-    destroy(last->next);
+  // Do we have this thing in the cache? If we don't, then we need to trigger
+  // loadCacheMiss()
+  auto resRecord = findCache(res);
+  if (resRecord == mResources.end()) {
+    if ((mCacheAccesses - mCacheHits) % 1 == 0) {
+      std::stringstream ss;
+      ss << "Replay cache miss. " << mCacheHits << " cache hits in "
+         << mCacheAccesses
+         << " accesses: " << (float)mCacheHits / (float)mCacheAccesses * 100.f
+         << " pc cache hit rate.";
+      GAPID_WARNING(ss.str().c_str());
```

Comment:
We probably don't want this to have as a warning, as this is expected behavior.
Also I would recommend using C-style formatting inside the GAPID_WARNING macro, that way the entire stringstream computation can get optimized out.